### ISSUE TITLE
REST API endpoint throwing wrong exception

### ISSUE THREAD

**Issue by paulespino** [magento/magento2#5191]
## **Steps to reproduce**
1. Install Magento from 2.1 branch.
2. Set up API integration
3. Send call to create Simple Product via /V1/products endpoint
4. Once simple has been added, create Configurable Product via V1/products endpoint
5. Add Simple Product created on **Step 3** via /V1/configurable-products/{sku}/child to the Configurable Product created on **Step 4**. 
## **Expected result**
1. 200 Response (return true) when adding the child product to the parent
## **Actual result**
1. [Example] Error message appears: "**Product has been already attached**".
   --- actual error ----
   {"message":"Product has been already attached","trace":"#0 [internal function]: Magento\ConfigurableProduct\Model\LinkManagement->addChild('576b124120360', '576b12351fb04')\n#1 \/var\/www\/staging.streetwear.la\/public_html\/app\/code\/Magento\/Webapi\/Controller\/Rest.php(307): call_user_func_array(Array, Array)\n#2 \/var\/www\/staging.streetwear.la\/public_html\/app\/code\/Magento\/Webapi\/Controller\/Rest.php(216): Magento\Webapi\Controller\Rest->processApiRequest()\n#3 \/var\/www\/staging.streetwear.la\/public_html\/var\/generation\/Magento\/Webapi\/Controller\/Rest\/Interceptor.php(37): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))\n#4 \/var\/www\/staging.streetwear.la\/public_html\/lib\/internal\/Magento\/Framework\/App\/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))\n#5 \/var\/www\/staging.streetwear.la\/public_html\/lib\/internal\/Magento\/Framework\/App\/Bootstrap.php(258): Magento\Framework\App\Http->launch()\n#6 \/var\/www\/staging.streetwear.la\/public_html\/index.php(44): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))\n#7 {main}"}

**In closing:**

I tested the the endpoint method addChild() located on line 95 on file:
**app/code/Magento/ConfigurableProduct/Model/LinkManagement.php**

Here is the method with notes:

```
public function addChild($sku, $childSku)
{
    $product = $this->productRepository->get($sku);
    $child = $this->productRepository->get($childSku);

    $childrenIds = array_values($this->configurableType->getChildrenIds($product->getId())[0]);

    if (in_array($child->getId(), $childrenIds)) {
        throw new StateException(__('Product has been already attached'));
    }

    $childrenIds[] = $child->getId();
    $product->setAssociatedProductIds($childrenIds);
    $product->save();
    return true;
}
```

This IF statement (in_array($child->getId(), $childrenIds)) fires if the ID of $childSku is returned by this line:

**$childrenIds = array_values($this->configurableType->getChildrenIds($product->getId())[0]);**

The weird part is that the exception that is executed here:

 **throw new StateException(__('Product has been already attached'));**

fires even after adding the FIRST Single Product to the Configurable and **that is what the BUG is.**

**Comment by veloraven** [comment]
@paulespino , thank you for your report.
I was able to reproduce this issue and get the result you're complaining, but only once and can not repeat it. Could you please provide more details about requests you used for products creation? Did you create configurable product with some other simples linked to it at once?

Anyway I created internal ticket (MAGETWO-55125) and linked it to this issue.

**Comment by paulespino** [comment]
Hi, 

Yes, the process was to create a child product via this endpoint: /V1/products

Once the child product was create, I can see it on the admin side. Then, I create the configurable product with the same endpoint (/v1/product). Once this is created, I add the child product to the configurable with this endpoint: /V1/configurable-products/{configurable_sku}/child

This last call to add the child is what responds with a "This child as already been added...". I ran thru the code that is creating this, and there is something weird happening. 

Please let me know if need more info to further discover this bug. I was able to replicate this issue over and over again...

**Comment by melnikovi** [comment]
@paulespino thank you for reporting the issue, unfortunately I can't reproduce it. When you get this error, do you see that simple product associated to configurable in admin? Could it be possible that in your code you execute two requests to associate simple product to configurable? Could you provide us code sample that you use to reproduce this issue?

Please see below endpoints and data I used to reproduce the issue.

POST http://magento.dev/pub/rest/default/V1/products

```
{
    "product": {
        "sku": "simple1",
        "name": "simple1",
        "attribute_set_id": 4,
        "price": 10,
        "status": 1,
        "visibility": 4,
        "type_id": "simple",
        "weight": 1,
        "custom_attributes": [
            {
                "attribute_code": "description",
                "value": "simple1"
            },
            {
                "attribute_code": "short_description",
                "value": "simple1"
            }
        ]
    }
}
```

POST http://magento.dev/pub/rest/default/V1/products

```
{
    "product": {
        "sku": "configurable1",
        "name": "configurable1",
        "attribute_set_id": 4,
        "price": 10,
        "status": 1,
        "visibility": 4,
        "type_id": "configurable",
        "weight": 1,
        "custom_attributes": [
            {
                "attribute_code": "description",
                "value": "configurable1"
            },
            {
                "attribute_code": "short_description",
                "value": "configurable1"
            }
        ]
    }
}
```

POST http://magento.dev/pub/rest/default/V1/configurable-products/configurable1/child

```
{
    "childSku": "simple1"
}
```

I also tried attaching multiple simple products to the same configurable and attaching the same simple product to two different configurable products. Works in both cases. When I attach simple product that is already associated, I get an error. But this is expected behaviour.

**Comment by melnikovi** [comment]
I'm closing this issue as we didn't get response in 2 weeks. If the problem still exists, please feel free to reopen

**Comment by melnikovi** [comment]
Sorry, overlooked that this is 2.1 branch. Reopening the issue

**Comment by cizgidekiadam** [comment]
Is there any update for this ?

**Comment by magento-engcom-team** [comment]
@paulespino, thank you for your report.
We've created internal ticket(s) MAGETWO-58507 to track progress on the issue.

**Comment by magento-engcom-team** [comment]
Hi @paulespino 
Could you please take a look at the comment above (https://github.com/magento/magento2/issues/5191#issuecomment-247182284) and let us know the result?

Thanks.

**Comment by bpicho** [comment]
Hello @magento-engcom-team as you requested, we tried to reproduce this bug, also using tips from (#5191 (comment)) but we could not reproduce described problem.

@paulespino Could you provide us more detailed information about your requests? 

We tested this on to branches 2.2-develop and 2.1-develop.

**Comment by magento-engcom-team** [comment]
Hi @paulespino. 

Thank you for your report and collaboration!

The related internal Jira ticket `MAGETWO-55125` was closed as `non-reproducible` in `2.3.x`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.
